docs: add proposal for TiDB_HOT_REGIONS_HISTORY (#27487)
pingcap/tidb: docs/design/2021-08-23-history-hot-regions.md

    + [Benchmark Tests](#benchmark-tests)
  * [Impacts & Risks](#impacts--risks)
  * [Investigation & Alternatives](#investigation--alternatives)

## Introduction

Create a new table `TIDB_HOT_REGIONS_HISTORY` in the `INFORMATION_SCHEMAT` schema to retrieve history hotspot regions stored by PD periodically.  

## Motivation or Background

TiDB has a memory table `TIDB_HOT_REGIONS` that provides information about hotspot regions. But it only shows the current hotspot information. This leads to the fact that when DBAs want to query historical hotspot details, they have no way to find the corresponding hotspot information.

According to the [documentation](https://docs.pingcap.com/tidb/stable/information-schema-tidb-hot-regions) for the current `TIDB_HOT_REGIONS` table,  it can only provides information about recent hotspot regions calculated by PD according to the heartbeat from tikv.  It is inconvenient to obtain hotspot regions of past time, and locate which store the region is. For ease of use, we can store extened hotspot region infomation in PD. The DBA can query hotspot regions within a specified period by such statement in TiDB:

```SQL

> SElECT * FROM information_schema.tidb_hot_regions_history WHERE update_time>='2019-04-01 00:00:00' and update_time<='2019-04-01 00:01:00';

Reviewer: add some other user case, such as group by store_id.

Author: add in new comment

  | PEER_ID     | bigint(21)  | YES  |      | NULL    |       | // new
  | IS_LEARNER  | tinyint(1)  | NO   |      | 0       |       | // new
  | IS_LEADER   | tinyint(1)  | NO   |      | 0       |       | // new
  | TYPE        | varchar(64) | YES  |      | NULL    |       |
  | HOT_DEGREE  | bigint(21)  | YES  |      | NULL    |       | // rename max_hot_degree to HOT_DEGREE
  | FLOW_BYTES  | double      | YES  |      | NULL    |       |
  | KEY_RATE    | double      | YES  |      | NULL    |       | // new
  | QUERY_RATE  | double      | YES  |      | NULL    |       | // new
  +-------------+-------------+------+------+---------+-------+
  | REGION_COUNT| bigint(21)  | YES  |      | NULL    |       | // deleted fields
  ```
  * Add `UPDATE_TIME` to support history.
  * Add `STORE_ID` ,  `PEER_ID`,  `IS_LAEDER`to track the machine of region.
  * Rename `MAX_HOT_DEGREE`  to `HOT_DEGREE` for precise meaning in history scenario.
  * Add `IS_LEARNER` and `IS_LEADER` to show role of regions.
  * Add `KEY_RATE` and `QUERY_RATE` for future expansion in hotspot determination dimensions.
  * Remove `REGION_COUNT` for disuse and repeat with `STORE_ID`.

> 2. Data size estimation 

Reviewer: move to PD storage part.

Author: move Data size estimation  to PD GC part

>     one record size: 1 * 4B(timestamp) + 1*1B(tinyint) + 6 * 8B(bitint) + 3 * 8B(double)  + 4 * 64B(varchar(64)) = 333B

Reviewer: why 4*64B?

Author: it's wrong, M of varchar(M) represents the maximum column length in characters (not bytes). The space occupied by a single character might differ for different character sets, form 1 to 4 bytes. it has ben addressed by new comment.
ref: [varchar type](https://docs.pingcap.com/tidb/stable/data-type-string#varchar-type)

>     Below table show data size per day and per month in 5,10,15 minutes record interval respectively given the maximum number of hotspot regions  is 1000:

Reviewer: ```suggestion
    Below table show data size per day and per month in 5,10,15 minutes record interval respectively given the maximum number of hotspot regions is 1000:
```

    | 333               | 10                  | 45.73059082            | 1371.917725              |
    | 333               | 15                  | 30.48706055            | 914.6118164              |

### PD
1. Timing write：

     The leader of PD will periodically encrypt  `START_KEY` and `END_KEY`  in data,and write hotspot region data into `LevelDB`.The write interval can be configured. The write fieldes are: `REGION_ID`, `TYPE`,  `HOT_DEGREE`, `FLOW_BYTES`, `KEY_RATE`, `QUERY_RATE`, `STORE_ID`, `PEER_ID`, `UPDATE_TIME`, `START_KEY`,  `END_KEY`.

2. Timing delete

   PD runs the delete check task periodically, and deletes the hotspot region data that exceeds the configured TTL time.

3. Pull interface

   PD query the data in `LevleDB` with filters pushed down, decodes the data and returns to TiDB.

4. New options

   There are two config options need to be add in PD’s `config.go`:

   * `HisHotRegionSaveInterval`:  time interval for pd to record hotspot region information, default: 10 minutes.
   * `HisHotRegionTTL`: maximum hold day for his hot region, default: 7 days. 0 means close.
   
5. GC

>      The amount of data stored for one month is as follows.

Reviewer: Some theoretical formulas can be added. For example, there are at most 1000 hotspot information, and each hotspot information is about how many bytes.
example: 
one month data write = 30(days) * 24(hours) * 6 (items/hours) * 100 B (each item size) * 1000 (max item numbers) = 411 MB. and the default save TTL is 7 days/

- Tracking Issue: https://github.com/pingcap/tidb/issues/25281

## Table of Contents

  * [Introduction](#introduction)
  * [Motivation or Background](#motivation-or-background)
  * [Detailed Design](#detailed-design)
    + [Existing TIDB HOT REGIONS](#existing-tidb-hot-regions)
    + [Table Header Design](#table-header-design)
    + [PD](#pd)
    + [TiDB](#tidb)
  * [Test Design](#test-design)
    + [Functional Tests](#functional-tests)
    + [Scenario Tests](#scenario-tests)
    + [Compatibility Tests](#compatibility-tests)
    + [Benchmark Tests](#benchmark-tests)
  * [Impacts & Risks](#impacts--risks)
  * [Investigation & Alternatives](#investigation--alternatives)

## Introduction

Create a new table `TIDB_HOT_REGIONS_HISTORY` in the `INFORMATION_SCHEMAT` schema to retrieve history hotspot regions stored by PD periodically.  

## Motivation or Background

TiDB has a memory table `TIDB_HOT_REGIONS` that provides information about hotspot regions. But it only shows the current hotspot information. This leads to the fact that when DBAs want to query historical hotspot details, they have no way to find the corresponding hotspot information.

> According to the [documentation](https://docs.pingcap.com/tidb/stable/information-schema-tidb-hot-regions) for the current `TIDB_HOT_REGIONS` table,  it can only provides information about recent hotspot regions calculated by PD according to the heartbeat from tikv.  It is inconvenient to obtain hotspot regions of past time, and locate which store the region is. For ease of use, we can store extened hotspot region infomation in PD. The following list some common user cases:

Reviewer: ```suggestion
According to the [documentation](https://docs.pingcap.com/tidb/stable/information-schema-tidb-hot-regions) for the current `TIDB_HOT_REGIONS` table,  it can only provides information about recent hotspot regions calculated by PD according to the heartbeat from tikv.  It is inconvenient to obtain hotspot regions of past time, and locate which store the region is. For ease of use, we can store extended hotspot region information in PD. The following list some common use cases:
```

Reviewer 2: change to proposal

| Field          | Type        | Null | Key  | Default | Extra |
+----------------+-------------+------+------+---------+-------+
| TABLE_ID       | bigint(21)  | YES  |      | NULL    |       |
| INDEX_ID       | bigint(21)  | YES  |      | NULL    |       |
| DB_NAME        | varchar(64) | YES  |      | NULL    |       |
| TABLE_NAME     | varchar(64) | YES  |      | NULL    |       |
| INDEX_NAME     | varchar(64) | YES  |      | NULL    |       |
| REGION_ID      | bigint(21)  | YES  |      | NULL    |       |
| TYPE           | varchar(64) | YES  |      | NULL    |       |
| MAX_HOT_DEGREE | bigint(21)  | YES  |      | NULL    |       |
| REGION_COUNT   | bigint(21)  | YES  |      | NULL    |       |
| FLOW_BYTES     | bigint(21)  | YES  |      | NULL    |       |
+----------------+-------------+------+------+---------+-------+
10 rows in set (0.00 sec)
```
There are two types of hotspot regions: `read` and `write`. The memory table retriever processes the following steps to fetch current hotspot regions from PD server:

1. TiDB send an HTTP request to the PD to obtain current hotspot regions. 
   
1. PD returns the following fields：

    ```go

>       // HotPeerStatShow records the hot region statistics for output

Reviewer: ```suggestion
    // HotPeerStatShow records the hot region statistics for output
```

2. Timing delete

   PD runs the delete check task periodically, and deletes the hotspot region data that exceeds the configured TTL time.

3. Pull interface

   PD query the data in `LevleDB` with filters pushed down, decodes the data and returns to TiDB.

4. New options

   There are two config options need to be add in PD’s `config.go`:

   * `HisHotRegionSaveInterval`:  time interval for pd to record hotspot region information, default: 10 minutes.
   * `HisHotRegionTTL`: maximum hold day for his hot region, default: 7 days. 0 means close.
   
5. GC

     * Data size estimation 

         one record size:

         > M of [varchar(M)](https://docs.pingcap.com/tidb/stable/data-type-string#varchar-type) represents the maximum column length in characters (not bytes). The space occupied by a single character might differ for different character sets, form 1 to 4 bytes. 

         minimum： 1 * 4B(timestamp) + 2 * 1B(tinyint) + 6 * 8B(bitint) + 3 * 8B(double)  + 4 * 64 * 1B(varchar(64)) = 334B

         maximum： 1 * 4B(timestamp) + 1 * 1B(tinyint) + 6 * 8B(bitint) + 3 * 8B(double)  + 4 * 64 * 4B(varchar(64)) = 1102B

>          Below table show data size per day and per month with 10 minutes interval, given the maximum number of hotspot regions is 1000, Note that varchar is used to store the name of table index, and its true size is generally smaller than the minimum case:

Reviewer: You should calculate the size of the serialized Region information stored on the PD side, not the field on the TiDB side to infer.

         | Record Length (B) | Time Interval (Min) | Data Size Per Day (MB) | Data Size Per Week (MB) | Data Size Per Month (MB) |
         | ----------------- | ------------------- | ---------------------- | ----------------------- | ------------------------ |
         | 334               | 10                  | 45.86791992            | 321.0754395             | 1376.037598              |
         | 1102              | 10                  | 151.3366699            | 1059.356689             | 4540.100098              |

>      * The amount of data stored for one month is as follows.

Reviewer: Is it the actual test result?

+----------------+-------------+------+------+---------+-------+
10 rows in set (0.00 sec)
```
There are two types of hotspot regions: `read` and `write`. The memory table retriever processes the following steps to fetch current hotspot regions from PD server:

1. TiDB send an HTTP request to the PD to obtain current hotspot regions. 
   
1. PD returns the following fields：

    ```go
    // HotPeerStatShow records the hot region statistics for output
    type HotPeerStatShow struct {
      StoreID        uint64    `json:"store_id"`
      RegionID       uint64    `json:"region_id"`
      HotDegree      int       `json:"hot_degree"`
      ByteRate       float64   `json:"flow_bytes"`
      KeyRate        float64   `json:"flow_keys"`
      QueryRate      float64   `json:"flow_query"`
      AntiCount      int       `json:"anti_count"`
      LastUpdateTime time.Time `json:"last_update_time"`
    }
    ```

1. After TiDB catch the response, it  fetch the `START_KEY` and `END_KEY` of the hot region from region cache or PD by `REGION_ID` to decode the corresponding schema information like：`DB_NAME`, `TABLE_NAME`,  `TABLE_ID`,  `INDEX_NAME`, `INDEX_ID`.

> 1. TiDB return the hotspot region row to the upper caller.

Reviewer 2: People who call the tidb_hot_regions_history table

> 1. TiDB return the hotspot region row to the upper caller.

Reviewer 2: If someone wants to view historical hot region information through the `tidb_hot_regions` table. the SQL statement is shown below.
`SELECT * FROM INFORMATION_SCHEMA.TIDB_HOT_REGIONS_HISTORY WHERE update_time >'2021-08-18 21:40:00' and update_time <'2021-09-19 00:00:00';`
when the statement is executed the above procedure is called to request information from PD.
The specific method and notes about the `tidb_hot_regions` table are shown in the following document
https://github.com/pingcap/docs/blob/0da6e2da0a5e3df10f3272436f6e4a4e484acfdc/information-schema/information-schema-tidb-hot-regions-history.md
